File: python/fabric/node/node.py
import asyncio
import logging
from typing import Dict, Optional, Callable, Any
from zenoh import Session, Subscriber, Publisher, Sample
from .interface import NodeInterface, NodeConfig
import json

# Set up logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)


class Node:
    def __init__(
        self, node_id: str, node_type: str, config: NodeConfig, session: Session
    ):
        self.id = node_id
        self.node_type = node_type
        self.config = config
        self.session = session
        self.interface: Optional[NodeInterface] = None
        self.publishers: Dict[str, Publisher] = {}
        self.subscribers: Dict[str, Subscriber] = {}

    async def run(self, cancel_token: asyncio.Event) -> None:
        logger.info(f"Node {self.id} starting...")
        try:
            await self.initialize()
            if self.interface:
                await self.interface.run(self, cancel_token)
        except Exception as e:
            logger.error(f"Error in node {self.id}: {str(e)}", exc_info=True)
        finally:
            logger.info(f"Node {self.id} shutting down...")
            await self.cleanup()
            logger.info(f"Node {self.id} cleaned up successfully")

    async def initialize(self) -> None:
        logger.info(f"Node {self.id} initialized with type {self.node_type}")
        try:
            await self.create_publisher(f"node/{self.id}/data")
            await self.create_subscriber(
                f"node/{self.id}/config", self.handle_config_update
            )
        except Exception as e:
            logger.error(f"Error initializing node {self.id}: {str(e)}", exc_info=True)
            raise

    async def cleanup(self) -> None:
        for publisher in self.publishers.values():
            publisher.undeclare()
        for subscriber in self.subscribers.values():
            subscriber.undeclare()

    async def create_publisher(self, topic: str) -> None:
        self.publishers[topic] = self.session.declare_publisher(topic)

    async def create_subscriber(
        self, topic: str, callback: Callable[[Sample], None]
    ) -> None:
        self.subscribers[topic] = self.session.declare_subscriber(topic, callback)

    async def publish(self, topic: str, data: Any) -> None:
        if topic in self.publishers:
            await self.publishers[topic].put(data)
        else:
            logger.warning(f"Attempted to publish to non-existent topic: {topic}")

    async def handle_config_update(self, sample: Sample) -> None:
        if self.interface:
            config_data = json.loads(sample.payload.decode())
            new_config = NodeConfig(**config_data)
            await self.interface.update_config(new_config)

    async def get_config(self) -> NodeConfig:
        if self.interface:
            return await self.interface.get_config()
        return self.config

    async def set_config(self, config: NodeConfig) -> None:
        self.config = config
        if self.interface:
            await self.interface.set_config(config)

    def get_type(self) -> str:
        return self.node_type

    async def handle_event(self, event: str, payload: dict) -> None:
        if self.interface:
            await self.interface.handle_event(event, payload)

File: rust/fabric/src/node/node.rs
use crate::error::{FabricError, Result};
use crate::node::generic::GenericNode;
use crate::node::interface::NodeData;
use crate::node::interface::{NodeConfig, NodeInterface};
use log::{debug, error, info, warn};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::mpsc;
use tokio::sync::{Mutex, RwLock};
use tokio::time::{interval, Duration};
use tokio_util::sync::CancellationToken;
use zenoh::prelude::r#async::*;

struct Publisher {
    topic: String,
    zenoh_publisher: zenoh::publication::Publisher<'static>,
}

pub struct Subscriber {
    topic: String,
    callback: Arc<Mutex<dyn Fn(Sample) + Send + Sync>>,
    zenoh_subscriber: zenoh::subscriber::Subscriber<'static, ()>,
}

#[derive(Clone)]
pub struct Node {
    id: String,
    node_type: String,
    config: Arc<RwLock<NodeConfig>>,
    session: Arc<Session>,
    interface: Arc<Mutex<Box<dyn NodeInterface + Send + Sync>>>,
    publishers: Arc<RwLock<HashMap<String, Publisher>>>,
    subscribers: Arc<RwLock<HashMap<String, Subscriber>>>,
    subscriber_tx: mpsc::Sender<Sample>,
}

impl Node {
    pub async fn new(
        id: String,
        node_type: String,
        config: NodeConfig,
        session: Arc<Session>,
        interface: Option<Box<dyn NodeInterface + Send + Sync>>,
    ) -> Result<Self> {
        let (subscriber_tx, subscriber_rx) = mpsc::channel(100);
        let interface = match interface {
            Some(interface) => interface,
            None => Box::new(GenericNode::new(config.clone())),
        };

        let node = Node {
            id,
            node_type,
            config: Arc::new(RwLock::new(config)),
            session,
            interface: Arc::new(Mutex::new(interface)),
            publishers: Arc::new(RwLock::new(HashMap::new())),
            subscribers: Arc::new(RwLock::new(HashMap::new())),
            subscriber_tx,
        };

        // Spawn a task to handle subscriber samples
        let node_clone = node.clone();
        tokio::spawn(async move {
            node_clone.handle_subscriber_samples(subscriber_rx).await;
        });

        Ok(node)
    }

    pub async fn run(&self, cancel: CancellationToken) -> Result<()> {
        info!("Starting node {}", self.id);

        let key_expr = format!("node/{}/config", self.id);
        let config_subscriber = self
            .session
            .declare_subscriber(&key_expr)
            .res()
            .await
            .map_err(|e| FabricError::ZenohError(e))?;

        // Initial status update
        self.update_status("online".to_string()).await?;

        // Spawn a task for periodic status updates
        let status_update_task = {
            let cancel_clone = cancel.clone();
            let self_clone = self.clone();
            tokio::spawn(async move {
                let mut interval = interval(Duration::from_millis(1000));
                loop {
                    tokio::select! {
                        _ = cancel_clone.cancelled() => {
                            break;
                        }
                        _ = interval.tick() => {
                            if let Err(e) = self_clone.update_status("online".to_string()).await {
                                warn!("Failed to update status for node {}: {:?}", self_clone.id, e);
                            }
                        }
                    }
                }
            })
        };

        loop {
            tokio::select! {
                _ = cancel.cancelled() => {
                    info!("Node {} received cancellation signal", self.id);
                    break;
                }
                sample = config_subscriber.recv_async() => {
                    match sample {
                        Ok(sample) => {
                            // TODO: Change this. Orchestrator publishes serialized JSON
                            let new_config: NodeConfig = serde_json::from_slice(sample.value.payload.contiguous().as_ref())
                                .map_err(|e| FabricError::SerdeJsonError(e))?;
                            info!("Node {} received new configuration: {:?}", self.id, new_config);
                            self.update_config(new_config).await?;
                        }
                        Err(e) => {
                            warn!("Error receiving configuration for node {}: {:?}", self.id, e);
                        }
                    }
                }
            }
        }

        // Wait for the status update task to complete
        status_update_task
            .await
            .map_err(|e| FabricError::Other(format!("Status update task error: {}", e)))?;

        info!("Node {} stopped", self.id);
        Ok(())
    }

    pub async fn update_config(&self, new_config: NodeConfig) -> Result<()> {
        self.interface
            .lock()
            .await
            .update_config(new_config.clone())
            .await;
        // Update the Node's config field
        let mut config = self.config.write().await;
        *config = new_config;
        Ok(())
    }

    pub async fn get_config(&self) -> NodeConfig {
        self.config.read().await.clone()
    }

    pub fn get_id(&self) -> &str {
        &self.id
    }

    pub fn get_type(&self) -> &str {
        &self.node_type
    }

    pub async fn get_interface(&self) -> Result<Arc<Mutex<Box<dyn NodeInterface + Send + Sync>>>> {
        Ok(self.interface.clone())
    }

    pub async fn set_interface(
        &mut self,
        interface: Box<dyn NodeInterface + Send + Sync>,
    ) -> Result<()> {
        *self.interface.lock().await = interface;
        Ok(())
    }

    pub async fn update_status(&self, status: String) -> Result<()> {
        let node_data = NodeData {
            node_id: self.id.clone(),
            node_type: self.node_type.clone(),
            status,
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map_err(|e| FabricError::Other(e.to_string()))?
                .as_secs(),
            metadata: None,
        };
        self.publish_node_status(&node_data).await
    }

    async fn publish_node_status(&self, node_data: &NodeData) -> Result<()> {
        let key_expr = format!("fabric/{}/status", self.id);
        let payload = serde_json::to_vec(node_data).map_err(|e| FabricError::SerdeJsonError(e))?;
        self.session
            .put(&key_expr, payload)
            .res()
            .await
            .map_err(|e| FabricError::ZenohError(e))?;
        debug!("Published status for node {}: {:?}", self.id, node_data);
        Ok(())
    }

    pub async fn create_publisher(&self, topic: String) -> Result<()> {
        let key_expr = topic.clone();
        let zenoh_publisher = self
            .session
            .declare_publisher(key_expr)
            .res()
            .await
            .map_err(|e| FabricError::ZenohError(e))?;

        let publisher = Publisher {
            topic: topic.clone(),
            zenoh_publisher,
        };
        debug!("Created publisher for topic: {}", publisher.topic.clone());

        self.publishers.write().await.insert(topic, publisher);
        Ok(())
    }

    pub async fn publish(&self, topic: &str, data: Vec<u8>) -> Result<()> {
        let publishers = self.publishers.read().await;
        if let Some(publisher) = publishers.get(topic) {
            publisher
                .zenoh_publisher
                .put(data)
                .res()
                .await
                .map_err(|e| FabricError::ZenohError(e))?;
            Ok(())
        } else {
            Err(FabricError::Other(format!(
                "Publisher not found for topic: {}",
                topic
            )))
        }
    }

    pub async fn create_subscriber(
        &self,
        topic: String,
        callback: Arc<Mutex<dyn Fn(Sample) + Send + Sync>>,
    ) -> Result<()> {
        let key_expr = topic.clone();
        let subscriber_tx = self.subscriber_tx.clone();
        let zenoh_subscriber = self
            .session
            .declare_subscriber(&key_expr)
            .callback(move |sample| {
                let tx = subscriber_tx.clone();
                tokio::spawn(async move {
                    if let Err(e) = tx.send(sample).await {
                        error!("Failed to send sample to handler: {:?}", e);
                    }
                });
            })
            .res()
            .await
            .map_err(|e| FabricError::ZenohError(e))?;

        let subscriber = Subscriber {
            topic: topic.clone(),
            callback,
            zenoh_subscriber,
        };

        debug!("Created subscriber for topic: {}", subscriber.topic);

        self.subscribers.write().await.insert(topic, subscriber);
        Ok(())
    }

    async fn handle_subscriber_samples(&self, mut rx: mpsc::Receiver<Sample>) {
        while let Some(sample) = rx.recv().await {
            let subscribers = self.subscribers.read().await;
            for subscriber in subscribers.values() {
                if subscriber
                    .zenoh_subscriber
                    .key_expr()
                    .intersects(sample.key_expr.as_keyexpr())
                {
                    let callback = subscriber.callback.lock().await;
                    callback(sample.clone());
                }
            }
        }
    }

    // Remove the old handle_subscriber_samples method
}
